### ISSUE TITLE
[Test Coverage] Implement Tests for Model Helpers

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#350]
## Problem
Model helpers in `src/Helpers/` have 0% test coverage.

## Files Requiring Tests (5 files)
- ModelHelper.cs
- NeuralNetworkHelper.cs
- OptimizerHelper.cs
- RegressionHelper.cs
- LayerHelper.cs

## Goal
Achieve 80%+ test coverage for model utility functions.

## Labels
test-coverage, helpers, models

**Comment by ooples** [comment]
# Issue #350: Junior Developer Implementation Guide
## Unit Tests for Model Helpers

---

## Overview

Create comprehensive unit tests for the model helper classes in `src/Helpers/` that assist with AI model operations.

**Files to Test**:
- `ModelHelper.cs` - Model creation and initialization
- `NeuralNetworkHelper.cs` - Neural network utilities
- `OptimizerHelper.cs` - Optimizer configuration
- `RegressionHelper.cs` - Regression model utilities
- `LayerHelper.cs` - Neural network layer utilities

**Target**: 0% → 80% test coverage

---

## ModelHelper Testing

### Key Methods to Test

1. **CreateDefaultModelData()**
   - Returns default empty X, Y, Predictions
   - Handles Matrix<T>/Vector<T> types
   - Handles Tensor<T> types
   - Throws exception for unsupported types

2. **CreateDefaultModel()**
   - Creates VectorModel for Matrix<T>/Vector<T>
   - Creates NeuralNetworkModel for Tensor<T>
   - Throws exception for unsupported combinations

3. **GetColumnVectors()**
   - Extracts columns from Matrix<T>
   - Extracts columns from Tensor<T>
   - Validates indices in range
   - Returns correct column values

4. **CreateRandomModelWithFeatures()**
   - Emphasizes active features
   - De-emphasizes inactive features
   - Works with vector models
   - Works with expression trees
   - Works with neural networks

5. **CreateRandomVectorModelWithFeatures()**
   - Sets meaningful weights for active features
   - Sets near-zero weights for inactive features
   - Returns valid VectorModel

6. **CreateRandomExpressionTreeWithFeatures()**
   - Uses only active feature indices
   - Respects max depth
   - Creates valid expression tree
   - Includes constants and variables

7. **CreateRandomNeuralNetworkWithFeatures()**
   - Creates valid architecture
   - Uses correct input size (total features)
   - Configures for active features

### Test Examples

```csharp
[TestClass]
public class ModelHelperTests
{
    [TestMethod]
    public void CreateDefaultModel_ForMatrixVector_ReturnsVectorModel()
    {
        // Act
        var model = ModelHelper<double, Matrix<double>, Vector<double>>.CreateDefaultModel();

        // Assert
        Assert.IsNotNull(model);
        Assert.IsInstanceOfType(model, typeof(VectorModel<double>));
    }

    [TestMethod]
    public void GetColumnVectors_WithValidIndices_ReturnsCorrectColumns()
    {
        // Arrange
        var matrix = new Matrix<double>(3, 4);
        for (int i = 0; i < 3; i++)
            for (int j = 0; j < 4; j++)
                matrix[i, j] = i * 4 + j;

        int[] indices = { 0, 2 };

        // Act
        var columns = ModelHelper<double, Matrix<double>, Vector<double>>
            .GetColumnVectors(matrix, indices);

        // Assert
        Assert.AreEqual(2, columns.Count);
        Assert.AreEqual(0.0, columns[0][0]);  // matrix[0,0]
        Assert.AreEqual(4.0, columns[0][1]);  // matrix[1,0]
        Assert.AreEqual(2.0, columns[1][0]);  // matrix[0,2]
        Assert.AreEqual(6.0, columns[1][1]);  // matrix[1,2]
    }

    [TestMethod]
    public void CreateRandomModelWithFeatures_EmphasizesActiveFeatures()
    {
        // Arrange
        int[] activeFeatures = { 0, 2 };  // Only features 0 and 2 active
        int totalFeatures = 5;

        // Act
        var model = ModelHelper<double, Matrix<double>, Vector<double>>
            .CreateRandomModelWithFeatures(activeFeatures, totalFeatures);

        // Assert
        var vectorModel = model as VectorModel<double>;
        Assert.IsNotNull(vectorModel);

        var coeffs = vectorModel.GetParameters();
        Assert.AreEqual(5, coeffs.Length);

        // Active features should have larger magnitude
        Assert.IsTrue(Math.Abs(coeffs[0]) > 0.01, "Active feature 0 should have meaningful weight");
        Assert.IsTrue(Math.Abs(coeffs[2]) > 0.01, "Active feature 2 should have meaningful weight");

        // Inactive features should have small magnitude
        Assert.IsTrue(Math.Abs(coeffs[1]) < 0.001, "Inactive feature 1 should have near-zero weight");
        Assert.IsTrue(Math.Abs(coeffs[3]) < 0.001, "Inactive feature 3 should have near-zero weight");
    }
}
```

---

## NeuralNetworkHelper Testing

### Key Methods

1. **CreateArchitecture()** - Build network architecture
2. **InitializeWeights()** - Initialize layer weights
3. **ValidateArchitecture()** - Check architecture validity
4. **GetLayerCount()** - Count network layers
5. **GetParameterCount()** - Count total parameters

### Test Checklist

- [ ] Architecture creation for different task types
- [ ] Weight initialization (Xavier, He, Random)
- [ ] Architecture validation (valid/invalid configs)
- [ ] Layer counting
- [ ] Parameter counting
- [ ] Input/output dimension validation

---

## OptimizerHelper Testing

### Key Methods

1. **CreateOptimizer()** - Factory method for optimizers
2. **ConfigureOptimizer()** - Set optimizer parameters
3. **GetDefaultLearningRate()** - Get recommended learning rate
4. **ValidateParameters()** - Check parameter validity

### Test Checklist

- [ ] Create SGD optimizer
- [ ] Create Adam optimizer
- [ ] Create RMSprop optimizer
- [ ] Configure learning rate
- [ ] Configure momentum
- [ ] Validate parameter ranges
- [ ] Default parameter values

---

## RegressionHelper Testing

### Key Methods

1. **CalculateYIntercept()** - Compute regression intercept
2. **CalculateCoefficients()** - Compute regression coefficients
3. **PredictValue()** - Make prediction
4. **CalculateR2()** - Calculate R-squared
5. **CalculateRMSE()** - Calculate root mean squared error

### Test Examples

```csharp
[TestMethod]
public void CalculateYIntercept_WithKnownRegression_ReturnsCorrectIntercept()
{
    // Arrange - y = 2x + 3
    var xMatrix = new Matrix<double>(3, 1);
    xMatrix[0, 0] = 1.0;
    xMatrix[1, 0] = 2.0;
    xMatrix[2, 0] = 3.0;

    var y = new Vector<double>(new[] { 5.0, 7.0, 9.0 });  // 2*x + 3
    var coefficients = new Vector<double>(new[] { 2.0 });  // Slope

    // Act
    double intercept = MathHelper.CalculateYIntercept(xMatrix, y, coefficients);

    // Assert
    Assert.AreEqual(3.0, intercept, 1e-10);
}
```

---

## LayerHelper Testing

### Key Methods

1. **CreateDenseLayer()** - Create fully connected layer
2. **CreateConvLayer()** - Create convolutional layer
3. **CreatePoolingLayer()** - Create pooling layer
4. **ValidateLayerConfig()** - Validate layer configuration
5. **GetOutputShape()** - Calculate output dimensions

### Test Checklist

- [ ] Dense layer creation
- [ ] Convolutional layer creation
- [ ] Pooling layer creation
- [ ] Layer configuration validation
- [ ] Output shape calculation for each layer type
- [ ] Activation function assignment
- [ ] Parameter initialization

---

## Test File Structure

```
tests/Helpers/
├── ModelHelperTests.cs
├── NeuralNetworkHelperTests.cs
├── OptimizerHelperTests.cs
├── RegressionHelperTests.cs
└── LayerHelperTests.cs
```

---

## Implementation Checklist

### ModelHelper
- [ ] 15+ tests covering all methods
- [ ] Test all input/output type combinations
- [ ] Test active feature emphasis
- [ ] Test random model creation
- [ ] Test column extraction

### NeuralNetworkHelper
- [ ] 12+ tests covering architecture
- [ ] Test different network types
- [ ] Test weight initialization methods
- [ ] Test validation logic

### OptimizerHelper
- [ ] 10+ tests covering optimizers
- [ ] Test all optimizer types (SGD, Adam, RMSprop, etc.)
- [ ] Test parameter configuration
- [ ] Test validation

### RegressionHelper
- [ ] 12+ tests covering regression
- [ ] Test intercept calculation (see MathHelper.CalculateYIntercept test)
- [ ] Test coefficient calculation
- [ ] Test metrics (R2, RMSE)

### LayerHelper
- [ ] 15+ tests covering layers
- [ ] Test each layer type
- [ ] Test configuration validation
- [ ] Test output shape calculation

---

## Success Criteria

- [ ] 5 test files created
- [ ] 64+ tests total (sum of all minimums)
- [ ] All tests passing
- [ ] Code coverage >= 75% for each helper
- [ ] Edge cases tested
- [ ] Integration between helpers tested

---

## Resources

- See ISSUE_349_JUNIOR_DEV_GUIDE.md for testing patterns
- See ISSUE_258_JUNIOR_DEV_GUIDE.md for null handling
- Use AAA pattern (Arrange-Act-Assert)

---

**Target**: Create comprehensive tests that ensure model helpers work correctly across all model types and configurations.
